README.md
[![NPM Version][npm-image]][npm-url]
[![NPM Downloads][downloads-image]][downloads-url]
[![Build Status][github-image]][github-url]
[![Test Coverage][coveralls-image]][coveralls-url]

# Description

**The most high resolution timing in NodeJs**

This module provide cross platform function rdtsc().

rdtsc() return counts the number of cycles since computer start

If you run in on Windows it will call C++ function __rdtsc() that call processor command [RDTSC (Read Time Stamp Counter)](https://en.wikipedia.org/wiki/Time_Stamp_Counter)

(rdtsc() - rdtsc()) is always < 0

Notice: you should not use parallel threads since you calc performance

# Install

```bash
npm i rdtsc -S
```
# Features

## RDTSC

```js
var { rdtsc } = require('rdtsc')
console.log(rdtsc()) // 3864063236708616 cycles
console.log(rdtsc() - rdtsc()) // -2971 cycles, minimum = -130 cycles
```

## Calc Performance

Accuracy: +/- 4 cycles

### Syntax
```js
/*

calcPerformance(testTimeMilliseconds, ...funcs)

result = {
    calcInfo: { 
        iterationCycles: 413.22476287663505,
        iterations: 5622917,
        funcsCount: 3,
        testTime: 1000 // milliseconds
    },
    cycles: [ 54n, 54n, 55n ], // is BigInt
    absoluteDiff: [ 0, 0 ], // cycles[i] - cycles[0]
    relativeDiff: undefined, // absoluteDiff[i] / absoluteDiff[0]
}

*/
```

### Examples

```js
const { calcPerformance } = require('rdtsc')

calcPerformance(1000, () => {}, () => {}, () => {})

/*
result = {
    calcInfo: { 
        iterationCycles: 413.22476287663505,
        iterations: 5622917,
        funcsCount: 3,
        testTime: 1000 // milliseconds
    },
    cycles: [ 54n, 54n, 54n ],
    absoluteDiff: [ 0, 0 ],
    relativeDiff: undefined,
}

OR

result = {
    ...
    cycles: [ 54n, 55n, 54n ],
    absoluteDiff: [ 1, 0 ],
    relativeDiff: [ 0 ],
}

OR

result = {
    ...
    cycles: [ 54n, 58n, 52n ],
    absoluteDiff: [ 4, -2 ],
    relativeDiff: [ -0.5 ],
}
*/

calcPerformance(1000, () => {}, () => {})

/*
result = {
    ...
    cycles: [ 54n, 54n ],
    absoluteDiff: [ 0 ],
    relativeDiff: undefined,
}
*/

calcPerformance(1000, () => {})

/*
result = {
    ...
    cycles: [ 54n ],
    absoluteDiff: undefined,
    relativeDiff: undefined,
}
*/

```

<!-- eslint-disable indent -->
```js
var { calcPerformance } = require('rdtsc')

var result = calcPerformance(
    1000,
    () => {

    },
    () => {
        Object.keys(Math)
    }
)

console.log('"Object.keys(Math)" min cycles =', result.absoluteDiff[0]) // about 20-40 cycles
```

## Thread Priority

Implemented only for Windows platform

### Examples

<!-- eslint-disable indent -->
```js

const { runInRealtimePriority, getThreadPriority, getProcessPriority } = require('rdtsc')

runInRealtimePriority(() => {
    console.log('getThreadPriority = ', getThreadPriority()) // === THREAD_PRIORITY_REALTIME
    console.log('getProcessPriority = ', getProcessPriority()) // === PROCESS_PRIORITY_REALTIME
})

```


<!-- eslint-disable indent -->
```js
const { setThreadPriority, getThreadPriority, isWin, THREAD_PRIORITY_REALTIME } = require('rdtsc')

if (isWin) {
    console.log(getThreadPriority()) // === THREAD_PRIORITY_NORMAL
} else {
    console.log(getThreadPriority()) // === undefined
}

var previousPriority = setThreadPriority(THREAD_PRIORITY_REALTIME)

try {
    // <your code>
} finally {
    setThreadPriority(previousPriority)
}
```

### Priorities list
```js
/*

const THREAD_PRIORITY_IDLE = -15
const THREAD_PRIORITY_LOWEST = -2
const THREAD_PRIORITY_BELOW_NORMAL = -1
const THREAD_PRIORITY_NORMAL = 0
const THREAD_PRIORITY_ABOVE_NORMAL = 1
const THREAD_PRIORITY_HIGHEST = 2
const THREAD_PRIORITY_REALTIME = 15 // THREAD_PRIORITY_REALTIME

*/
```

## Process Priority

Implemented only for Windows platform

### Example

<!-- eslint-disable indent -->
```js
const { setProcessPriority, getProcessPriority, isWin, PROCESS_PRIORITY_REALTIME } = require('rdtsc')

if (isWin) {
    console.log(getProcessPriority()) // === PROCESS_PRIORITY_NORMAL
} else {
    console.log(getProcessPriority()) // === undefined
}

var previousPriority = setProcessPriority(PROCESS_PRIORITY_REALTIME)

try {
    // <your code>
} finally {
    setProcessPriority(previousPriority)
}
```

### Priorities list
```js
/*

const PROCESS_PRIORITY_IDLE = 0x00000040 // IDLE_PRIORITY_CLASS
const PROCESS_PRIORITY_BELOW_NORMAL = 0x00004000 // BELOW_NORMAL_PRIORITY_CLASS
const PROCESS_PRIORITY_NORMAL = 0x00000020 // NORMAL_PRIORITY_CLASS
const PROCESS_PRIORITY_ABOVE_NORMAL = 0x00008000 // ABOVE_NORMAL_PRIORITY_CLASS
const PROCESS_PRIORITY_HIGHEST = 0x00000080 // HIGH_PRIORITY_CLASS
const PROCESS_PRIORITY_REALTIME = 0x00000100 // REALTIME_PRIORITY_CLASS

*/
```

# License

[Unlimited Free](LICENSE)

[npm-image]: https://img.shields.io/npm/v/rdtsc.svg
[npm-url]: https://npmjs.org/package/rdtsc
[downloads-image]: https://img.shields.io/npm/dm/rdtsc.svg
[downloads-url]: https://npmjs.org/package/rdtsc
[github-image]: https://github.com/NikolayMakhonin/nodejs-rdtsc/actions/workflows/test.yml/badge.svg
[github-url]: https://github.com/NikolayMakhonin/nodejs-rdtsc/actions
[coveralls-image]: https://coveralls.io/repos/github/NikolayMakhonin/nodejs-rdtsc/badge.svg
[coveralls-url]: https://coveralls.io/github/NikolayMakhonin/nodejs-rdtsc


HISTORY.md
3.0.0 / 2022-06-26
===================

* Big refactoring
* Added TypeScript definitions
* Added async functions
* Improved tests

2.0.0 / 2018-12-12
===================

* Support node v12.4.0
* Tested on node v10.9.0 - v12.10.0

1.3.1 / 2018-12-12
===================

  * calcPerformance function can evaluate performance of many functions

1.2.2 / 2018-12-03
===================

  * Improved calcPerformance function
  * Removed dependencies

1.2.1 / 2018-12-03
===================

  * Removed rdtsc0, rdtsc1 functions
  * Improved calcPerformance function

1.1.7 / 2018-12-03
===================

  * Changed minimal node version to 10.9.0

1.1.4 / 2018-12-03
===================

  * Improved code quality

1.1.2 / 2018-12-02
===================

  * Fix int64 type for support not windows platforms

1.1.1 / 2018-12-02
===================

  * Change license to CC0-1.0

1.1.0 / 2018-12-02
===================

  * Added function: rdtsc0, rdtsc1 for better quality resolution. Examlpe to use: rdtsc0();func();cycles=rdtsc1();
  * Improved calcPerformance function

1.0.6 / 2018-12-02
===================

  * Added function: runInRealtimePriority
  
1.0.3 / 2018-12-02
===================

  * Added functions: setProcessPriority, getProcessPriority
  * Improved calcPerformance function
  
1.0.1 / 2018-12-02
===================

  * Added functions: setThreadPriority, getThreadPriority
  * Improved calcPerformance function

src/index.js
const {
	rdtsc,
	init,
	mark0,
	mark1,
	minCycles,
	setThreadPriority,
	getThreadPriority,
	setProcessPriority,
	getProcessPriority,
	isWin
} = require('../build/Release/binding')

const THREAD_PRIORITY_IDLE = -15
const THREAD_PRIORITY_LOWEST = -2
const THREAD_PRIORITY_BELOW_NORMAL = -1
const THREAD_PRIORITY_NORMAL = 0
const THREAD_PRIORITY_ABOVE_NORMAL = 1
const THREAD_PRIORITY_HIGHEST = 2
const THREAD_PRIORITY_REALTIME = 15 // THREAD_PRIORITY_TIME_CRITICAL

const PROCESS_PRIORITY_IDLE = 0x00000040 // IDLE_PRIORITY_CLASS
const PROCESS_PRIORITY_BELOW_NORMAL = 0x00004000 // BELOW_NORMAL_PRIORITY_CLASS
const PROCESS_PRIORITY_NORMAL = 0x00000020 // NORMAL_PRIORITY_CLASS
const PROCESS_PRIORITY_ABOVE_NORMAL = 0x00008000 // ABOVE_NORMAL_PRIORITY_CLASS
const PROCESS_PRIORITY_HIGHEST = 0x00000080 // HIGH_PRIORITY_CLASS
const PROCESS_PRIORITY_REALTIME = 0x00000100 // REALTIME_PRIORITY_CLASS

const runInRealtimePriority = function (func) {
	if (!isWin) {
		return func()
	}

	const previousThreadPriority = getThreadPriority()
	const previousProcessPriority = getProcessPriority()

	try {
		setProcessPriority(PROCESS_PRIORITY_REALTIME)
		setThreadPriority(THREAD_PRIORITY_REALTIME)

		return func()
	} finally {
		setProcessPriority(previousProcessPriority)
		setThreadPriority(previousThreadPriority)
	}
}

// https://stackoverflow.com/a/39838385/5221762
const flatMap = (arr, callbackfn) =>
	arr.reduce((result, item) =>
		result.concat(callbackfn(item)), [])

const calcPerformance = function (testTimeMilliseconds, ...funcs) {
	return runInRealtimePriority(() => {
		const testTime = testTimeMilliseconds
		if (!testTime || testTime <= 0) {
			throw new Error(`testTime ${testTime} <= 0`)
		}

		const f = flatMap(
			funcs,
			o => {
				if (o == null) {
					return []
				}
				if (Array.isArray(o)) {
					return o
				}
				return [o]
			})
			.filter(o => {
				if (typeof o !== 'function') {
					throw new Error(`argument (${o}) is not a function`)
				}
				return true
			})

		const funcsCount = f.length

		if (!funcsCount) {
			throw new Error('functions count == 0')
		}

		const m0 = mark0
		const m1 = mark1
		const endTime = process.hrtime()
		endTime[0] += ~~(testTime / 1000)
		endTime[1] += testTime % 1000

		let i = 0
		let count = funcsCount
		init(funcsCount)
		const startCycles = rdtsc()
		do {
			const fn = f[i % funcsCount]
			m0()
			fn()
			m1()

			i++
			if (i >= count) {
				const time = process.hrtime(endTime)
				if (time[0] >= 0) {
					break
				}
				count = ~~Math.ceil(i * testTime / (testTime + time[0] * 1000 + time[1] / 1000000))
				count = (~~(count / funcsCount)) * funcsCount
			}
		} while (true)

		const cycles = minCycles()

		const absoluteDiff = funcsCount > 1
			? cycles.filter((o, i) => i).map(o => Number(o - cycles[0]))
			: undefined

		const relativeDiff = funcsCount > 2 && absoluteDiff[0] > 0
			? absoluteDiff.filter((o, i) => i).map(o => o / absoluteDiff[0])
			: undefined

		return {
			calcInfo: {
				iterationCycles: Number(rdtsc() - startCycles) / i,
				iterations: i,
				funcsCount,
				testTime
			},
			cycles,
			absoluteDiff,
			relativeDiff
		}
	})
}

module.exports = {
	rdtsc,
	calcPerformance,
	isWin,

	setThreadPriority, // since only for Windows
	getThreadPriority, // since only for Windows

	THREAD_PRIORITY_IDLE,
	THREAD_PRIORITY_LOWEST,
	THREAD_PRIORITY_BELOW_NORMAL,
	THREAD_PRIORITY_NORMAL,
	THREAD_PRIORITY_ABOVE_NORMAL,
	THREAD_PRIORITY_HIGHEST,
	THREAD_PRIORITY_REALTIME,

	setProcessPriority, // since only for Windows
	getProcessPriority, // since only for Windows

	PROCESS_PRIORITY_IDLE,
	PROCESS_PRIORITY_BELOW_NORMAL,
	PROCESS_PRIORITY_NORMAL,
	PROCESS_PRIORITY_ABOVE_NORMAL,
	PROCESS_PRIORITY_HIGHEST,
	PROCESS_PRIORITY_REALTIME,

	runInRealtimePriority // since only for Windows
}
